README.md
Digiquidus Explorer - 1.7.4
================

An open source block explorer written in node.js.

### See this fork in action

*  [DigitalNote XDN](https://xdn-explorer.com/)

*note: If you would like your instance mentioned here contact me*

### Updated Prereqs for this fork
* **LetsEncrypt** (if you choose https)
* **[ipstack.com](ipstack.com) - sign up for a free account and get API Key**

### **LetsEncrypt**
Only perform this step if you have a domain name linked and wish to use https (and this is reflected in yout settings file)
```
sudo apt-get -y install aptitude    
sudo aptitude -y install letsencrypt
sudo letsencrypt certonly --standalone --email email@example.com --agree-tos -d example.com -d www.example.com
   ``` 
   **PLEASE NOTE: Requirements for Certificate Auto-Renewal**
   1. Create Acme Challenge folders to serve AFTER you have finished the Explorer installation
   ```
   cd explorer/public
   sudo mkdir -p .well-known/acme-challenge
   sudo chmod 755 .well-known
   sudo chmod 755 .well-known/acme-challenge
   ```
   2. Update your ```/etc/letsencrypt/renewal/<domain>.conf ``` file with the following lines (replace ```"<user>"``` with your username):
   ```
   authenticator = webroot
   webroot_path = /<user>/explorer/public
   ``` 

### Explorer Build

Also check out [Beginners Guide for Iquidus Explorer Setup](https://gist.github.com/samqju/b9fc6c007f083e6429387051e24da1c3)

### Requires

    *  node.js >= 8.17.0 (12.14.0 is advised for updated dependencies)
    *  mongodb 4.2.x
    *  *coind

### Install MongoDB
https://docs.mongodb.com/manual/tutorial/install-mongodb-on-ubuntu/

    sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv 0C49F3730359A14518585931BC711F9BA15703C6
    echo "deb [ arch=amd64,arm64 ] http://repo.mongodb.org/apt/ubuntu xenial/mongodb-org/3.4 multiverse" | sudo tee /etc/apt/sources.list.d/mongodb-org-3.4.list
    sudo apt-get update
    sudo apt-get install -y mongodb-org
    sudo service mongod start

cd /var/log/mongodb
tail mongod.log
[initandlisten] waiting for connections on port <port>
Port 27017 by default

### Install Node.js

    sudo apt-get update
    sudo apt-get install nodejs nodejs-legacy -y
    sudo apt-get install npm

### Create database

Enter MongoDB cli:

    $ mongo

Create databse:

    > use explorerdb

Create user with read/write access:

    > db.createUser( { user: "digiquidus", pwd: "3xp!0reR", roles: [ "readWrite" ] } )

*Note: If you're using mongo shell 4.2.x, use the following to create your user:

    > db.addUser( { user: "username", pwd: "password", roles: [ "readWrite"] })

### Get the source

    git clone https://github.com/CryptoCoderz/digiquidus explorer
    sudo apt-get install libkrb5-dev

### Install node modules

    cd explorer && npm install --production
 
(If using root add:)
    
    --unsafe-perm=true --allow-root

### Configure

    cp ./settings.json.template ./settings.json

*Make required changes in settings.json*

### Start Explorer

    npm start

*note: mongod must be running to start the explorer*

As of version 1.4.0 the explorer defaults to cluster mode, forking an instance of its process to each cpu core. This results in increased performance and stability. Load balancing gets automatically taken care of and any instances that for some reason die, will be restarted automatically. For testing/development (or if you just wish to) a single instance can be launched with

    node --stack-size=10000 bin/instance

To stop the cluster you can use

    npm stop

### Syncing databases with the blockchain

sync.js (located in scripts/) is used for updating the local databases. This script must be called from the explorers root directory.

    Usage: node scripts/sync.js [database] [mode]

    database: (required)
    index [mode] Main index: coin info/stats, transactions & addresses
    market       Market data: summaries, orderbooks, trade history & chartdata

    mode: (required for index database only)
    update       Updates index from last sync to current block
    check        checks index for (and adds) any missing transactions/addresses
    reindex      Clears index then resyncs from genesis to current block

    notes:
    * 'current block' is the latest created block when script is executed.
    * The market database only supports (& defaults to) reindex mode.
    * If check mode finds missing data(ignoring new data since last sync),
      index_timeout in settings.json is set too low.


*It is recommended to have this script launched via a cronjob at 1+ min intervals.*

**crontab**

*Example crontab; update index every minute and market data every 2 minutes*

    */1 * * * * cd /path/to/explorer && /usr/bin/nodejs scripts/sync.js index update > /dev/null 2>&1
    */2 * * * * cd /path/to/explorer && /usr/bin/nodejs scripts/sync.js market > /dev/null 2>&1
    */5 * * * * cd /path/to/explorer && /usr/bin/nodejs scripts/peers.js > /dev/null 2>&1

### Wallet

Digiquidus Explorer is intended to be generic so it can be used with any wallet following the usual standards. The wallet must be running with atleast the following flags

    -daemon -txindex

### Donate

    BTC: 168hdKA3fkccPtkxnX8hBrsxNubvk4udJi
    XDN: JZp9893FMmrm1681bDuJBU7c6w11kyEY7D

### Known Issues

**script is already running.**

If you receive this message when launching the sync script either a) a sync is currently in progress, or b) a previous sync was killed before it completed. If you are certian a sync is not in progress remove the index.pid from the tmp folder in the explorer root directory.

    rm tmp/index.pid
    
I use script (launched by crontab):

```#!/bin/bash
   fname="/root/explorer/tmp/index.pid"
   echo "checking" `date` >> /root/explorer/tmp/scriptran.log 
   if [[ -f "$fname" ]];
   then
        pid=$(</root/explorer/tmp/index.pid)
        echo $pid
        if [ killall -0 $pid ]; then
                echo "script running"  `date` >> /root/explorer/tmp/script.log 
                exit 1	
        else
                rm $fname
		echo "removed" `date` >> /root/explorer/tmp/script.log 
        fi
   fi
   ```

**exceeding stack size**

    RangeError: Maximum call stack size exceeded

Nodes default stack size may be too small to index addresses with many tx's. If you experience the above error while running sync.js the stack size needs to be increased.

To determine the default setting run

    node --v8-options | grep -B0 -A1 stack_size

To run sync.js with a larger stack size launch with

    node --stack-size=[SIZE] scripts/sync.js index update

Where [SIZE] is an integer higher than the default.

*note: SIZE will depend on which blockchain you are using, you may need to play around a bit to find an optimal setting*

### License

Copyright (c) 2015, Digiquidus Technology  
Copyright (c) 2015, Luke Williams  
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.

* Neither the name of Digiquidus Technology nor the names of its
  contributors may be used to endorse or promote products derived from
  this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


CHANGELOG
1.7.4
* Updated themes to Bootstrap 4.5 by using latest Bootswatch themes
* Deleted unmaintained themes that were never ported to Bootstrap4
* New setting to optionally show a logo in the navbar header: headerlogo
* New settings to switch between default, dark and light boostswatch navbars: display.navbar_dark and display.navbar_light
* Added flag icons to network view.

1.7.3
* Fixes for Historical Address Balance - seems to be fastest method that works along-side parallel indexing
* New settings to optionally show market cap in header: show_market_cap and show_market_cap_over_price

1.7.2
* Index blocks in parallel - uses settings for task counts.
* Indexing is much faster (again).
* New settings to adjust parallel task counts: block_parallel_tasks

1.7.1
* Added "claim address" system - allows users to label addresses by signing messages
  
1.7.0
* Various updated dependencies & node version
* Switched from Jade to Pug
* Updated required mongo version

1.6.2
* Indexing is now MUCH faster.
* Added AddressTX model/collection - REQUIRES REINDEX
* Added balance history to Address TX History
* Added support for AJAX loading of Address TX History, using txcount setting for loading
* Stopped maximum TX count trimming TX's in AddressTX collection
* Removed tx_array from Address collection -> AddressTX
* Fix mismatching balances/sent/received, negative balances etc
* Added new file lock during database indexing: tmp/db_index.pid
* New setting to lock during indexing: lock_during_index
* Add setting to call bitcoin-core directly or use RPC during indexing

1.6.1
* fixed last_txs setting
* added hashrate_units setting
* added cryptsy support (markets)
* removed market update lockfile

1.6.0
* Fixed negative/incorrect balance due to sending to yourself (Issues #6, #8)
* Pages found via search now display full url 
* Markets backend overhaul/rewrite
* Added Yobit support (markets)
* Added Empoex support (markets)
* Layout redesign
* Layout data (coin supply, connections etc) now fetched via ajax
* Dynamic table data on main view (ajax, updates every 30 seconds)

1.5.2
* fix poloniex href (xCore)
* added links (xCore)
* coinbase balance fix (xCore)
* getaddress balance fix
* added getbalance api call
* finalized getlasttxs api call
* added footer logo

1.5.1
* added labels setting (address labels)

1.5.0
* added theme setting to settings.json
* added bootswatch themes
* new extended api call (getlasttxs)
* Index page now shows last x transactions instead of only last block
* added last_txs settings to settings.json
* mobile fixes
* OP_RETURN/type=nulldata crash fix
* PoS fixes
* search tx confirmation fix
* search tx blockcount fix
* address page redesign

1.4.1

* Added user and password to database settings
* Mongo authentication 
* Added alert (index not up-to-date)
* New richlist configuration settings
* added 'npm stop' to kill cluster
* Added logo setting

1.4.0

* Removed mintpal support (markets)
* Added poloniex support (markets)
* Cluster support
* Jasmine test scripts
* getaddress api fix (address not found)
* Base 2 rounding fix (values stored in satoshis)
* XDN fixes
* New current supply setting: BALANCES
* Sync fixes/optimizations
* New setting: show_sent_received

1.3.4

* Added qrcode api
* Added qrcode to address page

1.3.3

* Wealth Distribution stats & charts (richlist)
* New locale settings (richlist)
* Extended api
* getmoneysupply (api)
* getaddress (api)
* getdistribution (api)
* Negative zero balance fix (address)

1.3.2

* HeavyCoin support
* New heavy setting (heavy)
* New rewards page (heavy)
* Current coin supply (coinbase total sent, getinfo.moneysupply, getsupply (heavy))
* New supply setting (supply)
* New txcount setting
* Address txs segfault fix
* Indexing optimizations
* DOS address page fix

1.3.1

* Rich Lists/Top 100
* PoS fixes
* ui fixes
* coinbase address segfault fix
* performance tweaks

1.3.0

* merged blocknotify.js and markets.js into sync.js
* added update and check timeout settings
* local tx index
* input addresses/amounts
* unixtime format (now human friendly date/time)
* fixed genesis crash
* added genesis settings
* local address index
* new address view
* search by address

1.2.1

* updated locale (markets)
* updated index view
* updated tx view
* updated block view
* added optional seach to navbar
* added confirmations setting
* added screenshots

1.2.0

* added jqplot and datatables
* new markets page 
* added mintpal support (markets)
* added bittrex support (markets)
* new display setting (markets, twitter) 
* new locale entries (markets) 
* new settings (markets, api)
* added icons to nav (explorer, markets, api, loading, twitter)
* new twitter setting

1.1.0

* fixed navbar title (was not reading from settings)
* added locale support (and updated views)
* new display setting to toggle menu items (api)

1.0.0

* initial release


lib/settings.js
/**
* The Settings Module reads the settings out of settings.json and provides
* this information to the other modules
*/

var fs = require("fs");
var jsonminify = require("jsonminify");


//The app title, visible e.g. in the browser window
exports.title = "DigitalNote-Explorer";

//The url it will be accessed from
exports.address = "digitalnote.blockcrawlers.xyz";

//URL of site that is hosting explorer e.g. example.com (NOT www.example.com) - fot letsencrypt
exports.site_url = "example.com";

//Specify if site is http or https (Note: HTTPS requires domainname redirected and LetsEncrypt run)
exports.https_site = false;

// logo
exports.logo = "/images/logo.png";


//The app favicon fully specified url, visible e.g. in the browser window
exports.favicon = "favicon.ico";

//Theme
exports.theme = "Cyborg";

//The Port ep-lite should listen to
exports.port = process.env.PORT || 80;


//coin symbol, visible e.g. MAX, LTC, HVC
exports.symbol = "XDN";


//coin name, visible e.g. in the browser window
exports.coin = "DigitalNote";


//This setting is passed to MongoDB to set up the database
exports.dbsettings = {
  "user": "digiquidus",
  "password": "3xp!0reR",
  "database": "blockchaindb",
  "address" : "localhost",
  "port" : 27017
};


//This setting is passed to the wallet
exports.wallet = { "host" : "127.0.0.1",
  "port" : 18094,
  "user" : "digitalnoterpc",
  "pass" : "SomeOtherWorldlyPasswordGoesHere1"
};


//Locale file
exports.locale = "locale/en.json",


//Menu items to display
exports.display = {
  "api": true,
  "market": true,
  "twitter": false,
  "facebook": false,
  "googleplus": false,
  "youtube": false,
  "search": true,
  "richlist": true,
  "movement": true,
  "network": true,
  "mempool": true,
  "masternodes": true
};


//API view
exports.api = {
  "blockindex": 0,
  "blockhash": "00000d8e7d39218c4c02132e95a3896d46939b9b95624cf9dd2b0b794e6c216a",
  "txhash": "3b9d152cb1370d54d1ea30d5e334a83a41ca9403011495b8743a53d53423004a",
  "address": "dZ33rLDBTCmjQUD49FzdRERvcVmDhvfHuy",
};

// markets
exports.markets = {
  "coin": "XDN",
  "exchange": "BTC",
  "enabled": ['bittrex'],
  "default": "bittrex"
};

// richlist/top100 settings
exports.richlist = {
  "distribution": true,
  "received": true,
  "balance": true
};

exports.movement = {
  "min_amount": 100,
  "low_flag": 1000,
  "high_flag": 10000
},

//index
exports.index = {
  "show_hashrate": true,
  "difficulty": "Hybrid",
  "last_txs": 100
};

// twitter
exports.twitter = "digiquidus";
exports.facebook = "yourfacebookpage";
exports.googleplus = "yourgooglepluspage";
exports.youtube = "youryoutubechannel";

exports.confirmations = 6;

//timeouts
exports.update_timeout = 125;
exports.check_timeout = 250;


//genesis
exports.genesis_tx = "3b9d152cb1370d54d1ea30d5e334a83a41ca9403011495b8743a53d53423004a";
exports.genesis_block = "00000d8e7d39218c4c02132e95a3896d46939b9b95624cf9dd2b0b794e6c216a";

exports.heavy = false;
exports.txcount = 100;
exports.show_sent_received = true;
exports.supply = "COINBASE";
exports.nethash = "getnetworkhashps";
exports.nethash_units = "M";

//ipstack.com, get yourself a free account and API Key - for geolocation of peers
exports.ipstack_apikey = "AAAABBBBCCCCCDuckie"

exports.labels = {};

exports.reloadSettings = function reloadSettings() {
  // Discover where the settings file lives
  var settingsFilename = "settings.json";
  settingsFilename = "./" + settingsFilename;

  var settingsStr;
  try{
    //read the settings sync
    settingsStr = fs.readFileSync(settingsFilename).toString();
  } catch(e){
    console.warn('No settings file found. Continuing using defaults!');
  }

  // try to parse the settings
  var settings;
  try {
    if(settingsStr) {
      settingsStr = jsonminify(settingsStr).replace(",]","]").replace(",}","}");
      settings = JSON.parse(settingsStr);
    }
  }catch(e){
    console.error('There was an error processing your settings.json file: '+e.message);
    process.exit(1);
  }

  //loop trough the settings
  for(var i in settings)
  {
    //test if the setting start with a low character
    if(i.charAt(0).search("[a-z]") !== 0)
    {
      console.warn("Settings should start with a low character: '" + i + "'");
    }

    //we know this setting, so we overwrite it
    if(exports[i] !== undefined)
    {
      exports[i] = settings[i];
    }
    //this setting is unkown, output a warning and throw it away
    else
    {
      console.warn("Unknown Setting: '" + i + "'. This setting doesn't exist or it was removed");
    }
  }

};

// initially load settings
exports.reloadSettings();
